KALTURAPlayerSDK/WidevineClassicCDM.m

```diff
@@ -0,0 +1,246 @@
+        } else {
+            KPLogError(@"Error: downloaded file is not in the Documents directory.");
+            // will return nil
+        }
+    } else {
+        // Online file
+        assetPath = assetUri;
+    }
+    
+    return assetPath;
+}
+
++(void)dispatchAfterInit:(dispatch_block_t)block {
+    
+    // TODO: assuming initialization takes less than 200 msec. 
+    
+    if ([wvInitialized boolValue]) {
+        if ([NSThread isMainThread]) {
+            block();
+        } else {
+            dispatch_sync(dispatch_get_main_queue(), block);
+        }
+        return;
+    } else {
+        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, 200 * NSEC_PER_MSEC), dispatch_get_main_queue(), block);
+    }
+}
+
+static WViOsApiStatus widevineCallback(WViOsApiEvent event, NSDictionary *attributes) {
+    return [WidevineClassicCDM widevineCallbackWithEvent:event attr:attributes];
+}
+
++(WViOsApiStatus)widevineCallbackWithEvent:(WViOsApiEvent)event attr:(NSDictionary*)attributes {
+    
+    BOOL ignoreEvent = NO;
+    KCDMEventType cdmEvent = KCDMEvent_Null;
+    
+    switch (event) {
+        // Normal flow
+        case WViOsApiEvent_Initialized:
+            wvInitialized = [NSNumber numberWithBool:YES];
+            break;
+            
+        case WViOsApiEvent_Registered: break;
+        case WViOsApiEvent_EMMReceived: 
+            cdmEvent = KCDMEvent_LicenseAcquired;
+            break;
+            
+        case WViOsApiEvent_Playing:
+            cdmEvent = KCDMEvent_AssetCanPlay;
+            break;
+            
+        case WViOsApiEvent_Stopped: break;
+            
+        case WViOsApiEvent_QueryStatus:
+            cdmEvent = KCDMEvent_AssetStatus;
+            break;
+            
+        case WViOsApiEvent_EMMRemoved: break;
+        case WViOsApiEvent_Unregistered: break;
+        case WViOsApiEvent_Terminated: break;
+            
+        // Errors
+        case WViOsApiEvent_InitializeFailed:
+            wvInitialized = [NSNumber numberWithBool:NO];
+            break;
+            
+        case WViOsApiEvent_EMMFailed: break;
+        case WViOsApiEvent_PlayFailed: break;
+        case WViOsApiEvent_StoppingOnError: break;
+            
+        default:
+            // Other events are just informative
+            ignoreEvent = YES;
+            break;
+    }
+    
+    if (ignoreEvent) {
+        return WViOsApiStatus_OK;
+    }
+    
+    NSString* assetPath = attributes[WVAssetPathKey];
+    NSLog(@"widevineCallback: event=%@ asset='%@' attr=%@", NSStringFromWViOsApiEvent(event), assetPath, attributes);
+    
+    if (!assetPath) {
+        // Not an asset event
+        return WViOsApiStatus_OK;
+    }
+    
+    // TODO: also include relevant parsed data from attributes.
+    NSDictionary* data = @{
+                           @"ProviderSpecificData": attributes,
+                           @"ProviderSpecificEvent": NSStringFromWViOsApiEvent(event)
+                           };
```

Comment:
Did you check that attributes and NSStringFromWViOsApiEvent(event) is not nil?
Because if not the app will crash